Refactor dispatch structure so fallback code lives inline.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26719 Tests for fallback boxed dispatch.
* #26368 Add boxed fallback
* **#26367 Refactor dispatch structure so fallback code lives inline.**
* #26857 Change calling convention of ATenDispatch from getOp to callUnboxed.

This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

ezyang committed:
Refactor dispatch structure so fallback code lives inline.

This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/ATenDispatch.cpp:
```diff
@@ -7,37 +7,31 @@ ATenDispatch & globalATenDispatch() {
-        oss << toString(static_cast<TensorTypeId>(i));
-        first = false;
-      }
+void ATenOpTable::reportError(TensorTypeId tid) const {
```
this doesn't seem to report anything? It just discards the constructed string?

ezyang replied:
It raises an exception.

smessmer replied:
oh I somehow thought line 19 was ending this function already, missed that it continues

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang commented:
@pytorchbot retest this please

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dispatch structure so fallback code lives inline."


This is necessary for boxed fallback, as boxed fallback must
live inside the templated code.  Error reporting code never
has to be in templated code, so that stays in the C++ file.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17448556](https://our.internmc.facebook.com/intern/diff/D17448556)

[ghstack-poisoned]
